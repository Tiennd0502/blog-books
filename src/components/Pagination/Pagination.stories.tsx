import React from 'react';
import Storybook from '@storybook/react';

import Pagination from './index';

import '@styles/main.css';

export default {
  title: 'Components/Pagination',
  component: Pagination,
} as Storybook.ComponentMeta<typeof Pagination>;

const Template: Storybook.ComponentStory<typeof Pagination> = (args) => (
  <Pagination {...args} />
);

export const Default = Template.bind({});
Default.args = {
  count: 6,
  onChange: (page: number) => {
    console.log('Change page', page);
  },
};
