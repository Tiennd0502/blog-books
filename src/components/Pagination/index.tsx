import React, { memo, useMemo } from 'react';
import { Button, Flex, Box, Text, Spacer } from '@chakra-ui/react';
import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';

import styles from './styles.module.css';

type Props = {
  count: number;
  currentPage: number;
  onChange: (page: number) => void;
};

const Pagination: React.FC<Props> = ({ count, currentPage, onChange }) => {
  const PageList = useMemo(() => {
    const pageList: React.ReactNode[] = [];

    for (let i = 1; i <= count; i += 1) {
      pageList.push(
        <Button
          key={i}
          role="button"
          variant="unstyled"
          color={currentPage === i ? 'orange.500' : 'black'}
          className={styles.paginationItem}
          onClick={() => onChange(i)}
        >
          {i}
        </Button>
      );
    }
    return pageList;
  }, [count, currentPage, onChange]);

  return (
    <Flex
      role="list"
      align="center"
      width="100%"
      borderTopWidth="1px"
      borderColor="grey.500"
    >
      <Text>
        Page {currentPage} of {count}
      </Text>
      <Spacer />
      <Box>
        <Button
          className={styles.paginationItem}
          variant="unstyled"
          disabled={currentPage === 1}
          onClick={() => onChange(currentPage - 1)}
        >
          <ChevronLeftIcon />
          Previous
        </Button>

        {PageList}
        <Button
          className={styles.paginationItem}
          ml="8px"
          variant="unstyled"
          disabled={currentPage === count}
          onClick={() => onChange(currentPage + 1)}
        >
          Next <ChevronRightIcon />
        </Button>
      </Box>
    </Flex>
  );
};

export default memo(Pagination);
