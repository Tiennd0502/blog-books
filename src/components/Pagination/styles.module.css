
.paginationItem:hover {
  text-decoration: underline;
  color: var(--color-secondary);
}
