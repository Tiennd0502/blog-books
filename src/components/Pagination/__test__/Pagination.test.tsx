import { render, fireEvent, cleanup } from '@testing-library/react';

import Pagination from '../index';

describe('Pagination component', () => {
  const count = 5;
  const currentPage = 2;
  const onChange = jest.fn();

  afterEach(cleanup);
  test('Should render Pagination component', () => {
    const { getByRole } = render(
      <Pagination count={count} currentPage={currentPage} onChange={onChange} />
    );

    expect(getByRole('list')).toBeInTheDocument();
  });

  test('should simulate click change page', () => {
    const { getAllByRole } = render(
      <Pagination count={count} currentPage={currentPage} onChange={onChange} />
    );
    const buttonList = getAllByRole('button');

    fireEvent.click(buttonList[3]);

    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith(3);
  });

  test('should simulate click button previous', () => {
    const { getByText } = render(
      <Pagination count={count} currentPage={currentPage} onChange={onChange} />
    );
    const btnPrevious = getByText(/Previous/i);
    fireEvent.click(btnPrevious);

    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith(currentPage - 1);
  });

  test('should simulate click button next', () => {
    const { getByText } = render(
      <Pagination count={count} currentPage={currentPage} onChange={onChange} />
    );
    const btnNext = getByText(/Next/i);
    fireEvent.click(btnNext);

    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith(currentPage + 1);
  });

  test('Matches snapshot', () => {
    const { container } = render(
      <Pagination count={count} currentPage={currentPage} onChange={onChange} />
    );

    expect(container).toMatchSnapshot();
  });
});
