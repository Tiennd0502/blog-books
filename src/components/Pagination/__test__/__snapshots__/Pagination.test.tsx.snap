// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component Matches snapshot 1`] = `
<div>
  <div
    class="css-fne6jm"
    role="list"
  >
    <p
      class="chakra-text css-0"
    >
      Page 
      2
       of 
      5
    </p>
    <div
      class="css-17xejub"
    />
    <div
      class="css-0"
    >
      <button
        class="chakra-button paginationItem css-4xx2wk"
        type="button"
      >
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            fill="currentColor"
          />
        </svg>
        Previous
      </button>
      <button
        class="chakra-button paginationItem css-1n8hau1"
        role="button"
        type="button"
      >
        1
      </button>
      <button
        class="chakra-button paginationItem css-93kc5l"
        role="button"
        type="button"
      >
        2
      </button>
      <button
        class="chakra-button paginationItem css-1n8hau1"
        role="button"
        type="button"
      >
        3
      </button>
      <button
        class="chakra-button paginationItem css-1n8hau1"
        role="button"
        type="button"
      >
        4
      </button>
      <button
        class="chakra-button paginationItem css-1n8hau1"
        role="button"
        type="button"
      >
        5
      </button>
      <button
        class="chakra-button paginationItem css-1d6mcmw"
        type="button"
      >
        Next 
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
